File: python/sumBST.py
class BinaryTree:
    def __init__(self, value, left=None, right=None):
        self.value = value
        
        self.left = left
        self.right = right
        

class TreeInfo:
    def __init__(self, min, max, valid, sum=0, size=0, total=0):
        self.min = min
        self.max = max
        self.size = size
        self.sum = sum
        self.total = total
        self.valid = valid
        
        
# Time -> 0(N) Space -> 0(d) where d is the max height of the tree
def sumBsts(tree):
    return sumBst(tree).total
    
def sumBst(tree):
    if tree == None:
        return TreeInfo(min=float("inf"), max=float("-inf"), valid=True)

    leftTreeInfo = sumBst(tree.left)
    rightTreeInfo = sumBst(tree.right)
    
    
    valid = False
    if leftTreeInfo.valid and rightTreeInfo.valid:
        if tree.value > leftTreeInfo.max and tree.value <= rightTreeInfo.min:
            valid = True
            
    bstSum = 0
    size = 0
    minVal = min(tree.value, leftTreeInfo.min, rightTreeInfo.min)
    maxVal = max(tree.value, leftTreeInfo.max, rightTreeInfo.max)
    total = leftTreeInfo.total + rightTreeInfo.total
    
    if valid:
        bstSum = leftTreeInfo.sum + rightTreeInfo.sum + tree.value
        size = leftTreeInfo.size + rightTreeInfo.size + 1
        
        if size > 2:
            total = bstSum
    
    return TreeInfo(min=minVal, max=maxVal, valid=valid, sum=bstSum, size=size, total=total)

left = BinaryTree(1)
right = BinaryTree(3)
tree = BinaryTree(2, left=left, right=right)
print(sumBsts(tree))

File: go/sumBST/sumBST.go
package main

import (
	"fmt"
	"math"
)


type BinaryTree struct {
	Value int

	Left *BinaryTree
	Right *BinaryTree
}


type TreeInfo struct {
	size 	int
	bstsum 	int
	max 	int
	min 	int
	total	int
	isvalid bool
}


func SumBsts(tree *BinaryTree) int {
	treeInfo := Sum(tree)
	return treeInfo.total
}


// Time -> 0(N) Space -> 0(d) where d is the max height of the tree
func Sum(tree *BinaryTree) TreeInfo {
	if tree == nil {
		return TreeInfo{isvalid: true, min: math.MaxInt, max: math.MinInt}
	}

	leftTree := Sum(tree.Left)
	rightTree := Sum(tree.Right)

	validbst := tree.Value > leftTree.max && tree.Value <= rightTree.min
	valid := validbst && leftTree.isvalid && rightTree.isvalid

	maxVal := max(tree.Value, max(leftTree.max, rightTree.max))
	minVal := min(tree.Value, min(leftTree.min, rightTree.min))

	bstSum := 0
	bstSize := 0
	total := leftTree.total + rightTree.total

	if valid {
		bstSum = tree.Value + leftTree.bstsum + rightTree.bstsum
		bstSize = 1 + leftTree.size + rightTree.size

		if bstSize >= 3 {total = bstSum}
	}

	return TreeInfo{size: bstSize, bstsum: bstSum, max: maxVal, min: minVal, total: total, isvalid: valid}

} 

func min(x, y int) int {
	if x < y {return x}
	return y
}

func max(x, y int) int {
	if x > y {return x}
	return y
}

func main() {
	left := &BinaryTree{Value: 7}
	right := &BinaryTree{Value: 6, Left: left}
	tree := &BinaryTree{Value: 2, Right: right}
	fmt.Println(SumBsts(tree))
}